packages/angular_devkit/architect/src/api.ts

```diff
@@ -0,0 +1,230 @@
+ * be reconstructed across memory boundaries (it's not serializable but all internal information
+ * are).
+ */
+export interface BuilderRun {
+  /**
+   * Unique amongst runs. This is the same ID as the context generated for the run. It can be
+   * used to identify multiple unique runs. There is no guarantee that a run is a single output;
+   * a builder can rebuild on its own and will generate multiple outputs.
+   */
+  id: number;
+
+  /**
+   * The builder information.
+   */
+  info: BuilderInfo;
+
+  /**
+   * The next output from a builder. This is recommended when scheduling a builder and only being
+   * interested in the result of that single run, not of a watch-mode builder.
+   */
+  result: Promise<BuilderOutput>;
+
+  /**
+   * The output(s) from the builder. A builder can have multiple outputs.
+   * This always replay the last output when subscribed.
+   */
+  output: Observable<BuilderOutput>;
+
+  /**
+   * The progress report. A progress also contains an ID, which can be different than this run's
+   * ID (if the builder calls scheduleBuilder or scheduleTarget).
+   * This will always replay the last progress on new subscriptions.
+   */
+  progress: Observable<BuilderProgressReport>;
+
+  /**
+   * Stop the builder from running. Returns a promise that resolves when the builder is stopped.
+   * Some builders might not handle stopping properly and should have a timeout here.
+   */
+  stop(): Promise<void>;
+}
+
+/**
+ * The context received as a second argument in your builder.
+ */
+export interface BuilderContext {
+  /**
+   * Unique amongst contexts. Contexts instances are not guaranteed to be the same (but it could
+   * be the same context), and all the fields in a context could be the same, yet the builder's
+   * context could be different. This is the same ID as the corresponding run.
+   */
+  id: number;
+
+  /**
+   * The builder info that called your function. Since the builder info is from the builder.json
+   * (or the host), it could contain information that is different than expected.
+   */
+  builder: BuilderInfo;
+
+  /**
+   * A logger that appends messages to a log. This could be a separate interface or completely
+   * ignored. `console.log` could also be completely ignored.
+   */
+  logger: logging.LoggerApi;
+
+  /**
+   * The absolute workspace root of this run. This is a system path and will not be normalized;
+   * ie. on Windows it will starts with `C:\\` (or whatever drive).
+   */
+  workspaceRoot: string;
+
+  /**
+   * The current directory the user is in. This could be outside the workspace root. This is a
+   * system path and will not be normalized; ie. on Windows it will starts with `C:\\` (or
+   * whatever drive).
+   */
+  currentDirectory: string;
+
+  /**
+   * The target that was used to run this builder.
+   * Target is optional if a builder was ran using `scheduleBuilder()`.
+   */
+  target?: Target;
```

Comment:
What do you think about moving this to `input` (or rather combining them)?
This would allow for future _mutli-target_ builders that are aware of all their targets. Such a builder would receive a list of "target/inputs" and then act upon them as efficiently as possible.  The tslint builder for instance could be applicable here as this would allow the builder to generate one comprehensive output (especially important for certain output reporters such as XML; which is currently broken in this scenario).

Reply:
The `input` is not accessible to builders, only `context` (this interface) and `options`. Also, since a builder is always instantiated once per target and/or per builder name itself, I'm not sure that would make sense. And unless the builder is watching, one input => one context => one builder call, so theoretically that `target` member could change between two calls.

For a multi-target builder, I think we should extend this interface and (potentially) make it a separate class, passing in `targets[]` or something.

Reply:
Sorry I meant `options`.  As long as there would be a way to match the options with the targets in that scenario, I'm ok with that.  Also, I wouldn't make all builders accept an array just ones that would be created/marked as multi-target (`createMultitargetBuilder` maybe?).  Just didn't want to box ourselves into a corner moving forward.

Reply:
Options should not have stuff that cannot be validated by the builder. It’s whats in the options object in the workspace. We could do smart defaults like in schematics but considering smart defaults never really went anywhere I’d like to avoid it for now. 

I think a separate builder class would be the way to go.